Js2

Anclagen opened the pull request:
This is my current working version I feel I have delivered most of the functionality to some degree.

Questions:
- At the moment I am grabbing all the posts for the search, and feed, then sorting/filtering the data, should I change this? and use sort/sort-by strings on the fetch request instead and make more requests as the user scrolls or changes the sort parameters.
-  I have the search results display in a separate section to the post feed, I assume this still meets this criteria? "User can search the post content feed".
- Should I try to break down some of my function further?  Especially in reference to my [post_cards.mjs](https://github.com/Anclagen/SoSocial/blob/JS2/js/modules/render/post_cards.mjsl) where I have a very large function to build my posts and add all the relevant listeners and functionality, I have split some of the code for the forms, and replies into other files.
- Error reporting with try catch blocks should I be placing these in the primary fetch functions or further down the chain where I use function to apply this to the page, or is it ok to nest multiple try catches by adding one at each stage?

Anclagen committed:
added task, and project plan in readme

Anclagen committed:
updated sass and bootstrap

Anclagen committed:
create JS files for index and profile, updated links

Anclagen committed:
Update index.html

Anclagen committed:
Created validation functions

Anclagen committed:
Create basic fetch functions

basic fetch, as well as a headers+data fetch function if pagination becomes a thing this can be updated an modified further. MyOption class might be useful.

Anclagen committed:
Get/Set/Delete local storage functions

Anclagen committed:
Hidden class for form inputs

Anclagen committed:
Login/Register functions and API handler class

Login and Register fuctions added, also created a API handler class. The API handler can be expanded further with function to perform the required API actions.

Anclagen committed:
UpdateAPI function

This function checks for user info stored in local storage, if found it create the APIHandler class, if no user data is stored it redirects the user to the login page. I also added a query string for previous page maybe I will use this to bring the user back to the last page they were on.

Anclagen committed:
updated login html

Added extra inputs for signup, and error div for validation check responses.

Anclagen committed:
Added form validation and submission

I combined the login/sign up forms with a toggle function to change the forms validation to match.

Anclagen committed:
if statement to remove body for get requests

Anclagen committed:
added some handler functions to test

Anclagen committed:
re-arranging files

Anclagen committed:
removed test function

Anclagen committed:
added logout on login page

Anclagen committed:
added updateProfile to handler, fixed a URL

Anclagen committed:
refined regex for validation

Anclagen committed:
Update README.md

Anclagen committed:
adding documentation

Anclagen committed:
adding documentation and minor tweaks

Anclagen committed:
more documentation

Anclagen committed:
Merge pull request #6 from Anclagen/Authentication/Login/Registration

Login and SIgnup functionality created, Authentication handling with login response object created too.

Anclagen committed:
Update profile.mjs

Anclagen committed:
added default user/banner image + 404 image

Anclagen committed:
css update for various banner aspect ratios

Anclagen committed:
moving

Anclagen committed:
Basic post function for testing feed

Will update with more content, buttons and event listeners later.

Anclagen committed:
removed console log

Anclagen committed:
adding ids to containers

Anclagen committed:
added basic page rendering and query string

1. Querystring to display others users profile, default is own.
2. get profile data to render banner/avatar and name, default images provided.
3. Added posts to feed, need to filter for user.

Anclagen committed:
created accordion function and stylings

Anclagen committed:
change name

Anclagen committed:
fixing variable name in class

Anclagen committed:
validation for returned web image links

Help weed out the strings, and bad links out there!

Anclagen committed:
unused import

Anclagen committed:
function and class for posting comment

Anclagen committed:
update import/exports

Anclagen committed:
Update profile.html

Anclagen committed:
fill function/post feed filtered/post comment function

Anclagen committed:
lower case image validation, change default banner name

Anclagen committed:
profile qstring added, follow/unfollow function added

Anclagen committed:
create avatar for following/followers

Anclagen committed:
update import/export

Anclagen committed:
remove console logs

Anclagen committed:
edit profile form added, follower/following sections

Anclagen committed:
edit profile functions and follower data rendered

Anclagen committed:
fixed follower name overflow

Anclagen committed:
no content-type header for bodiless requests

Anclagen committed:
editing follower html setup

Anclagen committed:
novalidate on for email input preventing js error messages

Anclagen committed:
follow btn on other users profiles

broke down some function into smaller functions to call

Anclagen committed:
added post <template>

testing out using templates on my pages

Anclagen committed:
createPost using <template> function added

Anclagen committed:
changed 200 to ok, logout needs fixing

Anclagen committed:
Update _layout.scss

Anclagen committed:
Tidying/commenting/follower update on follow btn use

Anclagen committed:
Merge pull request #46 from Anclagen/Profile-Page

Profile Functionality, Merged
All basic needs met and additional requirement mostly completed.

Anclagen committed:
forgotten minor code changes

Anclagen committed:
offset added for getting next page

Anclagen committed:
not needed

Anclagen committed:
adding ids and containers for content

Anclagen committed:
updating imports

Anclagen committed:
try catch block added

Anclagen committed:
added post form, and got post feed rendering

Anclagen committed:
login => entry.html, fixed double stringify on login response

Anclagen committed:
Update index.mjs

Anclagen committed:
refactoring/rearranging code

Anclagen committed:
accordion now enable/disables hidden inputs.

Anclagen committed:
refactored comment posting, renamed some files

Anclagen committed:
updated comment posting/refactored some profile code

Anclagen committed:
added container accordion

Anclagen committed:
added reply to post feature

Anclagen committed:
tidy up/name change

Anclagen committed:
Create tidyup.jpg

Anclagen committed:
added height defining code to accordion

Anclagen committed:
added edit/delete/comment posts

Anclagen committed:
Refactored, Fixed bugs created by post page updates

Anclagen committed:
Merge pull request #50 from Anclagen/Posts-Page

Added posting/editing/commenting/deleting to posts page, and updated profile page to matchPosts Page

Anclagen committed:
Update index.html

Anclagen committed:
file restructuring

Anclagen committed:
Added search post functionality, and get all post method

Anclagen committed:
Update main.mjs

Anclagen committed:
Fixing bugs, with update post and following on profile

Anclagen committed:
Add sort and time filter, added loader

Anclagen committed:
Merge pull request #53 from Anclagen/Search-&-Filters

Search & filters added, further upgrades to API handler to get all posts

Anclagen committed:
added post modal to display specific post

Anclagen committed:
fixed bug

Anclagen committed:
making comment adds response to top of comments

Anclagen committed:
Create pngfind.com-cookie-monster-png-1790556.png

Anclagen committed:
Reply rendering, Reply tree filter and code documentation

Anclagen committed:
Merge pull request #54 from Anclagen/Post-Specific-Modal

Basic requirements met further functionality to be added during peer review and bug fixing if time permits

Anclagen committed:
Update README.md

Anclagen committed:
comments

Anclagen committed:
fixing validation, and register form

Anclagen committed:
commented and refactored login/signup code

Anclagen committed:
styles update

Anclagen committed:
Filled in user details and avatar, plus fix redirect errors

Anclagen committed:
redirect fix and current page updated on nav

Anclagen committed:
changed to accommodate login/signup changes

Anclagen committed:
post reply tree sorter, and reply render updates

Anclagen committed:
broke down post template, with edit/comment forms added

Anclagen committed:
Can reply to replies now

Anclagen committed:
sorting tag data to an array

Anclagen committed:
error reporter

Anclagen committed:
Added like/dislike buttons for posts

Anclagen committed:
post fetch and search use same data

No longer fetching two copies of the data from the API

Anclagen committed:
Search tags, and posted tags converted to array now

Anclagen committed:
Merge pull request #55 from Anclagen/Peer_Review_&_Bug_Fix

Peer Review, Bug Fix, & General Improvements

Anclagen committed:
Unused Image Purge

Anclagen committed:
show more posts on scroll listener added

Anclagen committed:
scrolling bug fix for filters

Anclagen committed:
Fixed multiple scroll listener instance after filter change.

Anclagen committed:
updating some names and cleaning up

martink-rsa commented:
Hi Anclagen, could you add me as a reviewer please?

martink-rsa commented on js/modules/render/loader.mjs:
```diff
@@ -0,0 +1,12 @@
+  container.innerHTML = `<div class="loader">
+                          <div class="outer-loader"></div>
+                          <div class="inner-loader"></div>
+                          <p>Getting products, please wait...</p>
```
This loader states we are getting products; however, there aren't products on the page.

Anclagen replied:
Loader now says "Getting posts, please wait..."

martink-rsa reviewed (changes requested):
Hi @Anclagen,

Great work on the work so far! 

You've done all of the assignment requirements. 

I've left one comment in the code for you to change. 

There is also feedback below for suggested changes to make. Great job once again.

## Formatting issues:
Your code has formatting issues. Please install Prettier, set it as your default formatter, and then set it to `Format on Save` in the settings. This will ensure that your code is always formatted and consistent. If you need assistance with this, please get in touch with one of the tutors.

## Hard-coded URLs:
You have hard-coded URLs, such as in the `register` function. This makes it challenging to maintain your code. If the server domain were to change, you would need to change this throughout your code, making room for errors. You should instead have this set as a variable in a file that gets imported; then you add the path onto it. E.g.

```js
export const SITE_URL = “https://nf-api.onrender.com/api/v1/social”;
```

Then you’d use this in your API calls and add the path you need e.g.
```js
Import { SITE_URL } from ‘./constans.mjs’;

const url = `${SITE_URL}/auth/register`;
```

## Don’t use shortened names for variables:
You have a variable `qstring` which is likely to mean query string. Variable naming like this can be problematic and cost more time than you save by typing less. A developer, including yourself at a later stage in time, will need to try to figure out what the `q` means in `qstring`. Rather write out the whole variable name so that your variable’s name is clear.

## Remove unused code before final submission:
There is unused code in your submission, such as code that is commented our and/or imports that aren’t being used. I understand that the code is not complete yet, however make sure it gets cleaned up before your final submission.

## Consider redirecting the user to the site once they have registered an account:
Consider redirecting the user once they have successfully registered an account. You aren’t required to do this, but if you’re looking for a challenge, then it’s something you could do to improve the UX. 

## `post_cards.mjs` functions abstracted:

Your `post_cards.mjs` file contains a lot of code and different functions, leading to 400 lines of code.

You should consider abstracting parts of this file into new files. This will make your code cleaner and more testable.

Please don't hesitate to contact the tutors if you need any assistance with the coursework. 

Kind regards,
Martin

martink-rsa commented:
Questions:

> 1. At the moment I am grabbing all the posts for the search, and feed, then sorting/filtering the data, should I change this? and use sort/sort-by strings on the fetch request instead and make more requests as the user scrolls or changes the sort parameters.

The way you have it is good. 

> 2. I have the search results display in a separate section to the post feed, I assume this still meets this criteria? "User can search the post content feed".

This still meets the criteria.

> 3. Should I try to break down some of my function further? Especially in reference to my [post_cards.mjs](https://github.com/Anclagen/SoSocial/blob/JS2/js/modules/render/post_cards.mjsl) where I have a very large function to build my posts and add all the relevant listeners and functionality, I have split some of the code for the forms, and replies into other files.

Yes, you should try abstracting some of this into new files. 

> 4. Error reporting with try catch blocks should I be placing these in the primary fetch functions or further down the chain where I use function to apply this to the page, or is it ok to nest multiple try catches by adding one at each stage?

Whenever you're dealing with the API will be a good place to use a `try catch` block. You can have them in your main API functions, but also in the individual functions if you're dealing with more API calls.

Anclagen committed:
Re-enabled prettier to format code

Anclagen committed:
Changed loader text

Anclagen committed:
broken up post function further

Anclagen committed:
added base url as a variable to fetches

Anclagen committed:
clearing up unused links and functions

Anclagen committed:
docs on divided functions

Anclagen committed:
User can logout, register now logs in user on success.

Anclagen committed:
Post Cards, the great abstraction completed

Render on scroll added to functionality.
post creator now split into 8 modules;
- edit and comment form.
- header, body, and footer.
- replies.
- options.
- react buttons.

Anclagen committed:
testing some search changes.

Anclagen committed:
css updates, more error handling

Anclagen committed:
forgot to change a variable name

Anclagen committed:
added loader to profile page, removed console logs

Anclagen committed:
added some more error feedback

Anclagen committed:
meta tags to pages

Anclagen committed:
Passing raw data for mutation, JSDocs updates, more error handling

Anclagen committed:
added if check for empty media, removed require on media input

Anclagen committed:
bug with updating a post twice

Anclagen committed:
removed unused code, added results counter to search

Anclagen committed:
clean up and documentation

Anclagen committed:
Merge pull request #61 from Anclagen/feedback-updates

Suggestions implemented, all the bugs I can find have been fixed, extra error handling added for those I might have missed.

Anclagen committed:
Created JS documentation, fixed some JSdoc errors

Anclagen committed:
Update README.md

Anclagen committed:
Update README.md
